The change itself looks fine to me, but it can lead to some weird diagnostics:
```c++
#define fold(x) (__builtin_constant_p(x) ? (x) : (x))

constexpr long fn(void) {
  return __builtin_bit_cast(long, fold((long)&fn));
}
static_assert(fn() == 1);
```
```console
array.cpp:714:15: error: static assertion expression is not an integral constant expression
  714 | static_assert(fn() == 1);
      |               ^~~~~~~~~
array.cpp:711:10: note: constexpr bit cast involving type 'long' is not yet supported
  711 |   return __builtin_bit_cast(long, fold((long)&fn));
      |          ^
array.cpp:714:15: note: in call to 'fn()'
  714 | static_assert(fn() == 1);
      |               ^~~~
```

but that's more a pre-existing problem and not introduced by this patch.

--- a/clang/lib/AST/ExprConstant.cpp
+++ b/clang/lib/AST/ExprConstant.cpp
@@ -7771,6 +7771,7 @@ class APValueToBufferConverter {
     case APValue::FixedPoint:
       // FIXME: We should support these.
 
+    case APValue::LValue:
     case APValue::Matrix:
     case APValue::Union:
     case APValue::MemberPointer:
@@ -7780,9 +7781,6 @@ class APValueToBufferConverter {
           << Ty;
       return false;
     }
-
-    case APValue::LValue:
-      llvm_unreachable("LValue subobject in bit_cast?");
     }
     llvm_unreachable("Unhandled APValue::ValueKind");
   }

--- a/clang/test/CodeGen/bitcast-based-lvalue.c
+++ b/clang/test/CodeGen/bitcast-based-lvalue.c
@@ -1,0 +1,7 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -emit-llvm -o - %s | FileCheck %s
+
+long fn(void) { return __builtin_bit_cast(long, (long)&fn); }
+
+// CHECK-LABEL: define{{.*}} i64 @fn()
+// CHECK: store i64 ptrtoint (ptr @fn to i64), ptr %ref.tmp, align 8
+// CHECK: ret i64 %{{.*}}

